### PR TITLE
Detect NuGet version and adjust behavior by per-version quirks

### PR DESCRIPTION
**Pull request by jmyersmsft**

### CODE REVIEW

**Comment by bryanmacfarlane** [comment]
Be aware we forked for 105 which is RC2 and RTM.  So port if you need in Dev15

**Comment by jmyersmsft** [comment]
@bryanmacfarlane will do.

File: Tasks/Common/nuget-task-common/NuGetQuirks.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    return { begin, beginIsInclusive: true, end, endIsInclusive: false };
+}
+
+function closedRange(begin: VersionInfoVersion, end: VersionInfoVersion): VersionRange {
+    return { begin, beginIsInclusive: true, end, endIsInclusive: true };
+}
+
+function versionIsInRange(version: VersionInfoVersion, range: VersionRange): boolean {
+    const beginComparison = VersionInfoVersion.compare(version, range.begin);
+    const endComparison = VersionInfoVersion.compare(range.end, version);
```
Comment (jmyersmsft) [inline Tasks/Common/nuget-task-common/NuGetQuirks.ts]: swap arguments and use < below

File: Tasks/Common/nuget-task-common/pe-parser/PEImageFile.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        await file.readAsync(buffer, 0, 2, 0);
+        const dosSignatureValue = buffer.readUInt16LE(0);
+        if (dosSignatureValue !== dosSignature) {
+            throw new PEParserError(
+                "invalidSignature",
+                "The DOS signature of the PE file does not match the expected value");
+        }
+
+        // read the pointer to the PE signature
+        await file.readAsync(buffer, 0, 4, 0x3c);
```
Comment (jmyersmsft) [inline Tasks/Common/nuget-task-common/pe-parser/PEImageFile.ts]: named constant

File: Tasks/Common/nuget-task-common/pe-parser/ResourceSection.ts [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    public getDataEntry(id: string | number): ResourceData {
+        return <ResourceData>this.entries.find(x => x.id === id && !(x.data instanceof ResourceDirectory)).data;
+    }
+
+    public getSubdirectory(id: string | number): ResourceDirectory {
+        return <ResourceDirectory>this.entries.find(x => x.id === id && x.data instanceof ResourceDirectory).data;
+    }
+}
+
+interface CookingResourceDirectory {
```
Comment (jmyersmsft) [inline Tasks/Common/nuget-task-common/pe-parser/ResourceSection.ts]: comment

File: Tasks/Common/nuget-task-common/pe-parser/ResourceSection.ts [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+export interface ResourceDirectoryEntry {
+    id: number | string;
+    data: ResourceData | ResourceDirectory;
+}
+
+export class ResourceDirectory {
+    public entries: ResourceDirectoryEntry[] = [];
+
+    public getDataEntry(id: string | number): ResourceData {
+        return <ResourceData>this.entries.find(x => x.id === id && !(x.data instanceof ResourceDirectory)).data;
```
Comment (jmyersmsft) [inline Tasks/Common/nuget-task-common/pe-parser/ResourceSection.ts]: don't crash if the entry isn't found

File: Tasks/Common/nuget-task-common/pe-parser/VersionResource.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    let end = offset;
+    while (end < buffer.length - 1) {
+        if (buffer.readInt16LE(end) === 0) {
+            return { value: buffer.toString("utf16le", offset, end), bytesRead: end - offset + 2 };
+        }
+
+        end += 2;
+    }
+
+    throw new Error("Unterminated string");
```
Comment (jmyersmsft) [inline Tasks/Common/nuget-task-common/pe-parser/VersionResource.ts]: use PEReaderError

**Comment by jmyersmsft** [comment]
I will do a tslint pass immediately after this PR

**Comment by jmyersmsft** [comment]
@zarenner @phil-hodgson 

I've added the following commits with PR feedback

https://github.com/Microsoft/vsts-tasks/pull/2336/commits/3aef40497935c2a0836d89eecbd4ae80aab47857

https://github.com/Microsoft/vsts-tasks/pull/2336/commits/8750d9e481b3ec58371741c0feb7d4e56538ab39?w=1

(the `?w=1` makes it ignore whitespace in the diff)

File: Tasks/Common/nuget-task-common/pe-parser/index.ts [hunk 6]
```diff
@@ -13,25 +13,36 @@ export async function getFileVersionInfoAsync(file: string | IReadableFile): Pro
...
+        if (!versionResourceBuffer) {
+            throw new PEParserError("noVersionResource", "No neutral-language version resource found in the file");
+        }
+
+        let versionResource = new VersionResource(versionResourceBuffer);
+        return versionResource.versionInfo;
+    } finally {
+        if (typeof file === "string") {
+            // we opened the file, so we're responsible for closing it
+            await readableFile.closeAsync();
```
Comment (phil-hodgson) [inline Tasks/Common/nuget-task-common/pe-parser/index.ts]: Add guard for readable file not being null? 
if (typeof file === "string" && readableFile) {....}

**Comment by phil-hodgson** [comment]
:shipit: (with comments)

**Pull request merged**